File: ocaml/src/p0059.ml
(* Project Euler: Problem 59 *)

(*
   ASCII
   reference: https://en.wikipedia.org/wiki/ASCII

   lower case characters:
   'a'..'z': 0x61(97d)..0x7a(122d)

   We'll need the following files to run this program.
   - https://projecteuler.net/project/resources/p059_cipher.txt   [encrypted data]
*)

open Core

let parse_data data =
  List.hd_exn data |> String.split ~on:',' |> List.map ~f:(fun s -> Int.of_string s)
;;

let calc_score ch =
  match ch with
  | n when n = 0x20 -> 3 (* SPC *)
  | n when n >= 0x41 && n <= 0x5A -> 5 (* 'A'..='Z' *)
  | n when n >= 0x61 && n <= 0x7A -> 3 (* 'a'..='z' *)
  | n when n >= 0x21 && n <= 0x7E -> 1 (* other ASCII characters *)
  | _ -> 0
;;

let compute str_lst =
  let cipher_text = parse_data str_lst in
  let rec loop max_score ans = function
    | [] -> ans
    | x :: xs ->
      let key = List.to_array x in
      let plain_text =
        List.mapi cipher_text ~f:(fun idx ch -> Int.bit_xor ch key.(idx mod 3))
      in
      let score = List.fold plain_text ~init:0 ~f:(fun acc ch -> acc + calc_score ch) in
      if score > max_score
      then loop score (List.sum (module Int) ~f:Fn.id plain_text) xs
      else loop max_score ans xs
  in
  loop
    0
    0
    (Euler.Util.permutation_with_repetition 3 (List.range 0x61 0x7A ~stop:`inclusive))
;;

let solve () =
  compute (Euler.Task.read_data "./src/assets/p059_cipher.txt") |> Int.to_string
;;

(* Test *)

let%test_unit "p059_cipher.txt" =
  [%test_eq: int] (compute (Euler.Task.read_file "./assets/p059_cipher.txt")) 129448
;;

File: rust/src/bin/p0059.rs
// Project Euler: Problem 59

euler::run_solver!(59);

static FILE_DATA: &str = include_str!("../../assets/0059_cipher.txt");

fn solve() -> String {
    compute(FILE_DATA).to_string()
}

fn compute(data: &str) -> i64 {
    use itertools::Itertools;

    let cipher_text = parse_data(data);
    let mut ans = 0_i64;
    let mut max_score = 0_i64;

    // 'a' = 0x61, 'z' = 0x7A
    for key in (0..3).map(|_| 0x61_u8..=0x7A).multi_cartesian_product() {
        let decrypted_text: Vec<_> = cipher_text
            .iter()
            .enumerate()
            .map(|(idx, n)| n ^ key[idx % 3])
            .collect();
        if decrypted_text
            .iter()
            .map(|&x| char::from(x)) // without this code, it is about 10% slower on my machine
            .all(|c| c.is_ascii_graphic() || c.is_ascii_whitespace())
        {
            let score = decrypted_text.iter().map(|ch| calc_score(*ch)).sum();
            if score > max_score {
                max_score = score;
                ans = decrypted_text.into_iter().map(i64::from).sum();
            }
        }
    }

    ans
}

fn parse_data(s: &str) -> Vec<u8> {
    s.split(',').map(|x| x.parse::<u8>().unwrap()).collect()
}

/*
  valuation rules:
    0x20 (space): 4
    0x41 - 0x5A (uppercase letters, 'A' - 'Z'): 3
    0x61 - 0x7A (lowercase letters, 'a' - 'z'): 5
    0x21 - 0x7E (printable characters except alphabet letters): 1
    others: 0
*/
fn calc_score(ch: u8) -> i64 {
    if ch == 0x20 {
        return 4;
    }
    if (0x41..=0x5A).contains(&ch) {
        return 3;
    }
    if (0x61..=0x7A).contains(&ch) {
        return 5;
    }
    if (0x21..=0x7E).contains(&ch) {
        return 1;
    }
    0
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0059() {
        assert_eq!(compute(super::FILE_DATA), 129448);
    }
}
